software/main.py

```diff
@@ -0,0 +1,127 @@
+from pyexpat.errors import XML_ERROR_UNKNOWN_ENCODING
+from types import *
+import image_processor
+import camera
+import motion
+import statemachine
+import cv2
+import numpy as np
+import time
+import referee
+
+# TODO if too much black around ball, ignore it
+# TODO create config file for constants
+# TODO make robot drive towards other basket if point basket is too close
+
+
+def main():
+    #---------
+    #config
+    competition = True
+    debug = True
+    #---------
+
+    cam = camera.RealsenseCamera(exposure = 100)
+    processor = image_processor.ImageProcessor(cam, debug=debug)
+    processor.start()
+    
+    omniRobot = motion.OmniMotionRobot()
+    stateMachine = statemachine.StateMachine(omniRobot)
+    stateMachine.throwIntoBlue = True
+    stateMachine.imageWidth = cam.rgb_width
+    stateMachine.imageHeight = cam.rgb_height
+    #stateMachine.currentState = statemachine.State.GO_TO_BASKET
+
+    if competition:
+        robotReferee = referee.Referee(ip="192.168.3.220", port="8111")
+        robotReferee.startReferee()
+        stateMachine.setState(statemachine.State.WAIT_REFEREE)
+
+    start = time.time()
+    fps = 0
+    frame = 0
+    useDepthImage = False
+    #frame counter
+    frame_cnt = 0
+    try:
+        while True:
+            #referee handling
+            if competition:
+                cmd = robotReferee.getCommand()
+                if (cmd != None):
+                    if cmd[0] == 'START':
+                        #stateMachine.currentState = statemachine.State.GO_TO_BASKET
+                        stateMachine.currentState = statemachine.State.FIND_BALL
+                        if cmd[1] == 'blue':
+                            stateMachine.throwIntoBlue = True
+                        else:
+                            stateMachine.throwIntoBlue = False
+
+                    elif cmd[0] == 'STOP':
+                        stateMachine.currentState = statemachine.State.WAIT_REFEREE
+            
+            
+            # has argument aligned_depth that enables depth frame to color frame alignment. Costs performance
+            processedData = processor.process_frame(aligned_depth=useDepthImage)
+            #updating state machine data
+            stateMachine.setImageData(processedData)
+
+            useDepthImage = False
+
+            if stateMachine.currentState == statemachine.State.FIND_BALL:
```

Comment:
Why have this if block? It sets the state state to the value of the current state? Maybe set the depth boolean if the state is one of the two that currently trigger it.

Reply:
The setState() method was written when we didn't understand that the statemachine can't use blocking methods. Replaced with executeState() method. Main loop is much cleaner now